frontend/src/pages/AccountSend.vue

```diff
@@ -590,20 +943,110 @@ function useSendForm() {
+        if (!Sends.value[i].receiver || !Sends.value[i].token || !Sends.value[i].amount)
+          throw new Error('please complete the form');
+      }
+      if (!signer.value) throw new Error('wallet not connected');
+      if (!umbra.value) throw new Error('Umbra instance not configured');
+
+      // Verify the recipient ID is valid. (This throws if public keys could not be found. This check is also
+      // done in the Umbra class `send` method, but we do it here to throw before the user pays for a token approval.
+      // This should usually be caught by the isValidId rule anyway, but is here again as a safety check)
+      const ethersProvider = provider.value as Provider;
+      for (let i = 0; i < Sends.value.length; i++) {
+        await umbraUtils.lookupRecipient(Sends.value[i].receiver as string, ethersProvider);
+      }
+      // Ensure user has enough balance. We re-fetch token balances in case amounts changed since wallet was connected.
+      // This does not account for gas fees, but this gets us close enough and we delegate that to the wallet
+      // await getTokenBalances();
+
+      for (let i = 0; i < Sends.value.length; i++) {
+        const token: TokenInfoExtended | null | undefined = Sends.value[i].token;
+        if (!token) throw new Error('please complete the form');
+
+        const { address: tokenAddress, decimals } = token;
+
+        const tokenAmount = parseUnits(Sends.value[i].amount, decimals);
+
+        if (tokenAddress === NATIVE_TOKEN.value.address) {
+          // Sending the native token, so check that user has balance of: amount being sent + toll
+          const requiredAmount = tokenAmount.add(toll.value);
+          if (requiredAmount.gt(balances.value[tokenAddress])) throw new Error('amount exceeds balance');
+        } else {
+          // Sending other tokens, so we need to check both separately
+          const nativeTokenErrorMsg = `${NATIVE_TOKEN.value.symbol} Send.umbra-fee-exceeds-balance`;
+          if (toll.value.gt(balances.value[NATIVE_TOKEN.value.address])) throw new Error(nativeTokenErrorMsg);
+          if (tokenAmount.gt(balances.value[tokenAddress])) throw new Error('Send.amount-exceeds-balance');
+        }
+
+        // If token, get approval when required
+        isSending.value = true;
+        if (token.symbol !== NATIVE_TOKEN.value.symbol) {
+          // Check allowance
+          const tokenContract = new Contract(token.address, ERC20_ABI, signer.value);
+          const batchSendAddress = umbra.value.batchSendContract!.address;
+          const allowance = (await tokenContract.allowance(userAddress.value, batchSendAddress)) as BigNumber;
+          // If insufficient allowance, get approval
+          if (tokenAmount.gt(allowance)) {
+            const approveTx = await tokenContract.approve(batchSendAddress, MaxUint256);
+            void txNotify(approveTx.hash as string, ethersProvider);
+            await approveTx.wait();
```

Comment:
Lots of awaits in loops
- the allowance checks should be pulled out into a map + Promise.all
- we can send off all approval txs WITHOUT calling `await approveTx.wait()`
- then we can await all of them at at the end for better UX: Right now if a user is doing a batch send with 2 tokens the flow is approve tx > wait for it to confirm > approve tx > wait for it to confirm > batch send. But it can be approve tx > approve tx > wait for both to confirm > batch send

Reply:
How does this look? https://github.com/ScopeLift/umbra-protocol/pull/491/commits/66644c31227edaa495faf3c895537973a62933b4

Reply:
Really great! Two small comments
- I think we need `promises.push(Promise.resolve(Zero))` instead of `promises.push(Zero)`. The latter isn't pushing a promise, whereas the former is pushing one that will immediately resolve. If the code is working, perhaps `Promise.all` is smart enough to handle the current code, though I haven't tested
- `batchSendAddress` can be defined as a constant outside of the loops instead of redeclaring the same value in all loop iterations

Reply:
- Pushing a promise def makes more sense. 
- Pulled `batchSendAddress` out of the loop
https://github.com/ScopeLift/umbra-protocol/pull/491/commits/38f23b03cc53a6ea369aba89182e6bf3b7fa9c43